### ISSUE TITLE
Oracle JAVA 15, Records - UnsupportedOperation can't get field offset on

### ISSUE THREAD

**Issue by zdenek-jonas** [microstream-one/microstream#164]
I cannot use Java Records on Oracle java 15.0.1

Just use this project in github:
https://github.com/johny2000uwb/microstream-records
and change in pom.xml line: 67
```         
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.1</version>
                <configuration>
                    <release>14</release> <!-- <release>13/14/15</release> -->
                    <compilerArgs>--enable-preview</compilerArgs>
                </configuration>
            </plugin>

```
to 

```
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.1</version>
                <configuration>
                    <release>15</release> <!-- <release>13/14/15</release> -->
                    <compilerArgs>--enable-preview</compilerArgs>
                </configuration>
            </plugin>
```
and call `mvn test`

Code:
```
public record PersonRecord(String firstName, String lastName) {

}
```
```
package test.microstream;

import one.microstream.storage.types.EmbeddedStorage;
import one.microstream.storage.types.EmbeddedStorageManager;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;
import test.microstream.data.PersonRecord;

import java.nio.file.Path;

public class RecordTest {

    @TempDir
    Path location;


    @Test
    public void saveRecordTest() {
        PersonRecord personRecord = new PersonRecord("Maria", "Lukasova");

        EmbeddedStorageManager storage = EmbeddedStorage.start(personRecord, location);
        storage.shutdown();

        PersonRecord secondRecord = new PersonRecord("Kamila", "Pazourkova");
        storage = EmbeddedStorage.start(secondRecord, location);

        Assertions.assertEquals("Maria", secondRecord.firstName());

    }

}
```
pom.xml:
```
<?xml version="1.0" encoding="UTF-8"?>
<project xmlns="http://maven.apache.org/POM/4.0.0"
         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
    <artifactId>records</artifactId>
    <groupId>one.microstream</groupId>
    <version>0.0.1-SNAPSHOT</version>
    <modelVersion>4.0.0</modelVersion>

    <properties>
        <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
        <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
        <microstream.version>04.00.00-MS-GA</microstream.version>
    </properties>

    <repositories>
        <repository>
            <id>microstream-releases</id>
            <url>https://repo.microstream.one/repository/maven-public/</url>
        </repository>
    </repositories>

    <dependencies>
        <dependency>
            <groupId>one.microstream</groupId>
            <artifactId>storage.embedded</artifactId>
            <version>${microstream.version}</version>
        </dependency>
        <dependency>
            <groupId>one.microstream</groupId>
            <artifactId>storage.embedded.configuration</artifactId>
            <version>${microstream.version}</version>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-api</artifactId>
            <version>5.4.2</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-params</artifactId>
            <version>5.4.2</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.junit.jupiter</groupId>
            <artifactId>junit-jupiter-engine</artifactId>
            <version>5.4.2</version>
            <scope>test</scope>
        </dependency>
        <dependency>
            <groupId>org.testcontainers</groupId>
            <artifactId>junit-jupiter</artifactId>
            <version>1.15.0</version>
            <scope>test</scope>
        </dependency>
    </dependencies>

    <build>
        <plugins>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-compiler-plugin</artifactId>
                <version>3.8.1</version>
                <configuration>
                    <release>15</release> <!-- <release>13/14/15</release> -->
                    <compilerArgs>--enable-preview</compilerArgs>
                </configuration>
            </plugin>
            <plugin>
                <groupId>org.apache.maven.plugins</groupId>
                <artifactId>maven-surefire-plugin</artifactId>
                <version>2.22.2</version>
                <configuration>
                    <argLine>--enable-preview</argLine>
                </configuration>
            </plugin>
        </plugins>
    </build>


</project>

```

Some Information about it:
https://mail.openjdk.java.net/pipermail/amber-dev/2020-June/006241.html

**Comment by zdenek-jonas** [comment]
Unfortunately, the problem is a little more complicated. The implementation in Java 15 is different from Java 14.
https://openjdk.java.net/jeps/395
For example: I can define my constructor now:
```
public record PersonRecord(String firstName, String lastName) {
    
    public PersonRecord(String firstName, String lastName) {
        this.firstName = firstName;
        this.lastName = firstName + lastName;
    }

    public PersonRecord(String firstName) {
        this(firstName, "default");
    }
}

```
So i can modify the data in Constructor. (Potential security problem when we use it to create instance in Microstream)

**Comment by tm-ms** [comment]
If records don't handle as normal classees reflection-wise, then the current state of MicroStream is that Records are not supported. This is not just a "bug" in handling records, but records are not yet supported at all.
To add support, it has to be researched, first, how records work. Their memory model, their reflective behavior, etc.

But from what I saw so far, this might not be so easy or maybe not even possible at all.
If the JVM forbids write accesses to field declared as final in records, then there is no way of generically creating an instance of the particular type and fill its fields with the required data. This would be a fundamental problem in the JVM, not just some missing function in MicroStream.
I bet EVERY other framework that generically creates and fills instances (i.e. EVERY serialization, persistence, generic something middleware framework) will have the same problem.
From what I can see at the moment, especially with the changes between Java 14 and 15, they screwed up the implementation of Records pretty bad and will have to fix that first before records become usable in a generic way.

I don't have any experience or understanding of records. Someone else should look into this.

Until the problem is resolved (probably first on the JDK level with a new Java version, THEN on our end), I suggest the official statement is "MicroStream does currently not support records".

**Comment by fh-ms** [comment]
They prohibited reflective access to record's fields in JDK 15. But there are RecordComponents as a substitute. 
It would be doable to write a generic type handler for records, but with a bigger effort. Since this must be done with Java >=14, we need to create an own module for this, and provide a special type handler creator for records.

**Comment by fh-ms** [comment]
Records are still in preview. They will be final in JDK 16 which corresponds to our release 5.0. Therefor this issue will be postponed.
Documentation has been updated accordingly.

**Comment by fh-ms** [comment]
JDK 16 is released. 

I investigated records again. 
The only visible difference to normal classes is, reflection-wise, the prohibited access to record field offsets.
They copied `sun.misc.Unsafe `to `jdk.internal.misc.Unsafe,` and delegated most method calls one by one. Except e.g. `objectFieldOffset(Field)`, which contains a simple if statement, which throws an exception when called for record fields. `jdk.internal.misc.Unsafe` is not accessible from outside, neither normally nor by reflection. 
 I don't get the deeper meaning behind this move, neither do other serialization framework developers, which caused an upset and issues in the official JDK tracker, but none of them are fixed, until now.

This means for us, the only way to access record field offsets, is by unlocking access to `jdk.internal.misc.Unsafe`, which can be done with following VM parameter:
`--add-exports java.base/jdk.internal.misc=ALL-UNNAMED`

I implemented a workaround for record field offset access, but in order to work, the described VM parameter has to be set.

This is all we can do right now. Documentation has to state this clearly.

The only alternative to this solution would be creating record instances by reflection. But then user-code is beeing executed at deserialization, and therefor no viable substitute.
